cEP-0021.md

```diff
@@ -0,0 +1,379 @@
+
+This proposal is to introduce the Profiling interface so that bear writer will
+have the ability to jump directly into Bear code and move through it using a
+debugger interface and Profile the Bear code to create a Better Performant
+bears.
+
+## Proposed Change
+
+Bear writer can invoke Profiler through *CLI*:
+
+- By passing custom settings through *CLI*
+  `coala -b Bear Mybear --files file --profile-bears -"True,...."`,
+- By using coala default settings provided by coala
+  `coala -b Bear Mybear --files file --profile-bears`,the default are to strip
+  the directory (`strip_dirs()`) and sort the stats of profiler result by
+  cummulative time (`cumtime`) and display the profiler result on *CLI*
+
+Bear writer can invoke Profiler by using `coafile` i.e. `profile_bears = XXX`.
+
+- The profiling will start by profiling the `run()` method of the bears.
+  This will load the settings passed to it, and then run the Profiler with the
+  passed settings on `run()` method of Bear.
+
+- The advantage of profiling just the `run()` method of the bear is that,
+  bear writer doesn't have to spend time on loading the files,
+  collecting the settings, etc. as this is all done by coala itself, they just
+  have to Profile the `run()` method of Bear to create a Better PerformantBear.
+
+After invoking the profiler on the run method of a bear it will first save
+all the profile output on a local temp file which further user by
+`pstats module` to profile results and formats them in a way Bear
+writer wants to format it.
+
+> Currently default format is to strip the directory (`strip_dirs()`) and sort
+> the stats of profiler result by cummulative time (`cumtime`) and display the
+> profiler result on *CLI*.
+>
+> ```python
+> pstats.strip_dirs().sort_stats('cumtime')
+> ```
+
+After formatting the profile output Bear writer will have an option to
+either display the formatted ouput on terminal or dump the formatted result
+in a specified file.
+
+```
+coala -b Bear Mybear --files <file> --profile-bears "<filename or file-path>"
+```
+
+Bear writer will also have a option to direclty dump the profile output
+to a file which can be further use to create a Sunburn Diagram for better
+estimation of profiler stats with the help of external library like `snakeviz`.
+
+```
+coala -b Bear Mybear --files <file> --profile-bears "True,dump"
+```
+
+# Implementation Detail
+
+When user invoke the profiler with either `CLI` or by `coafile`
+then `execute` method of Bear Base Class will call the `run_bear_from_section`
+method
+
+```python
+def run_bear_from_section(self, args, kwargs):
+    ............
+    profile_bears_results, args, kwargs = self.setup_profiler(
+        *args, **kwargs)
+    return profile_bears_results if profile_bears_results else self.run(
+        *args, **kwargs)
+```
+
+which futher call the `setup_profiler()` method, and `setup_profiler()` method
+first create a `Profile`  ( name `prof`) then enable the profiler on `run()`
```

Comment:
break the sentence, what do you mean by create a Profile? say something like turn on the profiler or create an instance of the class